### PR TITLE
Kernel/USB: Add support for async & interrupt transfers

### PR DESCRIPTION
**Pull request by b14ckcat**
Depends on: https://github.com/SerenityOS/serenity/pull/15631 and https://github.com/SerenityOS/serenity/pull/15529 (their commits are present in this PR for convenience/ease of testing and will be removed once these PRs or their equivalent are merged)

The first commit here adds support for interrupt transfers, as well as async bulk transfers. The implementation here uses a separate polling kernel task to check a list of active async transfers at a set interval of 1ms (the period of a single UHCI frame, chosen specifically because interrupt transfers are only executed by the hardware controller once per frame and don't take advantage of bandwidth reclamation), invoking the user-provided callback on those that are complete as needed. 

A nicer way of implementing this instead of polling (courtesy of @ADKaster) would be to use a binary semaphore that is counted up within an ISR that is invoked when interrupt transfers are complete (possible through the interrupt-on-complete flag within the UHCI transfer data structures), and then having the async task wait on this semaphore repeatedly. It doesn't seem like there is a semaphore type present in the kernel yet though, with the closest being the wait queue, which would also work just as well but would be a bit more clunky and less expressive. Right now the async poll process seems to use around 0.8% CPU consistently (on my system anyway) which is pretty undesirable, so this is probably a good thing to look into.

The second commit isn't meant to be merged, but instead provides an example (and method of testing) of interrupt transfers working in case anyone is interested in taking a look. The actual USB device drivers will be written to target the USB driver probe branch here: https://github.com/Quaker762/serenity/commits/usb_probe

Comments and feedback are appreciated :)

### CODE REVIEW

**Comment by Quaker762** [comment]
Commit f58c567c143786d7dcb9d0f150d00a0c5101e614 and 84e9c182c87e935a4eef6a5def06d4a6a2a85a78 need to be dropped from this PR

File: Kernel/Bus/USB/USBPipe.h
Comment (Quaker762) [inline Kernel/Bus/USB/USBPipe.h:21]: Maybe

```suggestion
using usb_async_callback = Function<void(Transfer* transfer)>
```
instead?

**Review by Quaker762 — CHANGES_REQUESTED** [review]
This is looking really good. Excellent work on getting the USB Driver going in the way you have, it looks great.

**Comment by b14ckcat** [comment]
> Commit [f58c567](https://github.com/SerenityOS/serenity/commit/f58c567c143786d7dcb9d0f150d00a0c5101e614) and [84e9c18](https://github.com/SerenityOS/serenity/commit/84e9c182c87e935a4eef6a5def06d4a6a2a85a78) need to be dropped from this PR

Agreed, will get rid of them when I undraft this PR after all of it's prerequisites are merged

**Comment by b14ckcat** [comment]
I reduced the CPU time used by the async poll process by having it sleep as long as it can (limited by the fastest required poll interval) before running again. That way if there are no async transfers queued it will amount to negligible overhead, and in other cases it will only run as often as it needs to. This probably introduces some jitter but I would imagine its not a huge deal.

File: Kernel/Bus/USB/UHCI/UHCIController.cpp
Comment (ADKaster) [inline Kernel/Bus/USB/UHCI/UHCIController.cpp:537]: Is this trying to bin the transfer based on the log base 2 of the millisecond interval? It's not quite clear whether we really need a loop here but my refactoring brain isn't finding a cleaner solution using some fancy bit manip methods.

File: Kernel/Bus/USB/UHCI/UHCIController.cpp
Comment (ADKaster) [inline Kernel/Bus/USB/UHCI/UHCIController.cpp:539]: move() the transfer here perhaps? This should be `TRY(try_append()` at the very least, which then makes it obvious that it's another possible allocation under a spinlock (due to vector resize) :/

Perhaps this should be a FixedArray with a size set at init? At least we should apply a bound to this vector/queue somehow before appending

File: Kernel/Bus/USB/UHCI/UHCIController.cpp
Comment (ADKaster) [inline Kernel/Bus/USB/UHCI/UHCIController.cpp:530]: this lock makes me nervous. You are doing multiple allocations here with it held. Kmalloc with a spinlock held is muy no bueno. at least move it below the create_transfer_queue call.

File: Kernel/Bus/USB/UHCI/UHCIController.cpp
Comment (ADKaster) [inline Kernel/Bus/USB/UHCI/UHCIController.cpp:621]: Instead of manually unlocking, add another scope

```
{
    SpinlockLocker locker(m_lock);
    for (...) {
    }
}
sleep()
```

File: Kernel/Bus/USB/USBPipe.cpp
Comment (ADKaster) [inline Kernel/Bus/USB/USBPipe.cpp:140]: The name here sounds like it could be trying to interrupt an in-progress transfer.

Perhaps `create_input_transfer_from_interupt` or similar?

File: Kernel/Bus/USB/USBPipe.h
Comment (ADKaster) [inline Kernel/Bus/USB/USBPipe.h:21]: Types should be PascalCase. `USBAsyncCallback`

File: Kernel/Bus/USB/UHCI/UHCIController.cpp
Comment (ADKaster) [inline Kernel/Bus/USB/UHCI/UHCIController.cpp:400]: Pre-existing since I see this code was just moved around, but passing two uninitialized pointers by address sets off my "This is C" alarm.

At the very least can we explicitly init those guys to nullptr? I would prefer if we could not use out-params like this and return `ErrorOr<SomeTinyStructWithTwoPointers>` rather than `ErrorOr<void>`

File: Kernel/Bus/USB/UHCI/UHCIController.cpp
Comment (ADKaster) [inline Kernel/Bus/USB/UHCI/UHCIController.cpp:413]: This explicit free feels like we're not leveraging RAII enough in this subsystem. I hope there's a matching free somewhere else after the transfer is completed, or in the destructor of the transfer queue?

**Review by ADKaster — CHANGES_REQUESTED** [review]
This seems like a good start at doing these things async, but there's a few things that stand out to me. Especially regarding allocations that may or may not be hidden.

File: Kernel/Bus/USB/UHCI/UHCIController.cpp
Comment (ADKaster) [inline Kernel/Bus/USB/UHCI/UHCIController.cpp:551]: Ah something else. You are removing an item from this list while iterating it. Which... is "ok" since you return right after doing so, but generally feels awkward. This is also a `find_if` algorithm. Refactoring to use AK::find_if (or a similar member on vector? I don't remember if there is one) would make the intent here clearer.

Comment (ADKaster) [reply Kernel/Bus/USB/UHCI/UHCIController.cpp:551]: Or even remove_if, if we have one.

File: Kernel/Bus/USB/UHCI/UHCIController.cpp
Comment (ADKaster) [inline Kernel/Bus/USB/UHCI/UHCIController.cpp:612]: micro-optimization, but you could could consider pre-calculating the min poll interval when appending to the queue and avoid having to re-calculate it every time around the loop.

e.g.

```
void add_transfer(Transfer* t, u16 poll_interval_ms)
{
    // setup some stuff
    
    Locker lock(m_lock);
    if (m_queue->min_poll_interval_ms > poll_interval_ms)
       m_queue->min_poll_interval_ms = poll_interval_ms;
    m_queue->add(t);
}
```
And then re-calculate the min after removing the item or items that have the min poll interval.

File: Kernel/Bus/USB/UHCI/UHCIController.cpp
Comment (ADKaster) [inline Kernel/Bus/USB/UHCI/UHCIController.cpp:622]: I think the only thing missing to make this "Semaphore at home" is a way to wake this thread up when appending to the queue. Just an unconditional `async_poll_thread->unblock()` from the "add transfer" method should do the trick. The no-args version, since that will call `Thread::unblock(u8 signal)` with a signal of 0, which will not dispatch any signals, just remove any thread blockers, namely the SleepBlocker from Thread::sleep().

File: Kernel/Bus/USB/UHCI/UHCIController.cpp
Comment (b14ckcat) [reply Kernel/Bus/USB/UHCI/UHCIController.cpp:537]: Yeah that's pretty much what it's doing. Basically we need to round it to a power of two so we can link it in to the closest schedule queue (the power of two part is only really required for high-speed interrupt endpoints & isochronous endpoints, but it's a good enough approximation and makes things easier for full/low speed interrupt transfers anyway, I will add a comment for this). We could use log2 here but it would involve float casts & arithmetic which is kind of awkward.

File: Kernel/Bus/USB/UHCI/UHCIController.cpp
Comment (b14ckcat) [reply Kernel/Bus/USB/UHCI/UHCIController.cpp:539]: Yeah the Vector could be replaced with FixedArray<AsyncTransferHandle>(MAXIMUM_NUMBER_OF_QHS) then we can avoid dynamic memory altogether.

File: Kernel/Bus/USB/UHCI/UHCIController.cpp
Comment (b14ckcat) [reply Kernel/Bus/USB/UHCI/UHCIController.cpp:400]: Agreed it would be nice to refactor this.

File: Kernel/Bus/USB/UHCI/UHCIController.cpp
Comment (b14ckcat) [reply Kernel/Bus/USB/UHCI/UHCIController.cpp:413]: Yeah this would be a good application of RAII, all of the cleanup is the responsibility of cancel_async_transfer currently (for async transfers).

**Comment by b14ckcat** [comment]
@ADKaster Ok, I've hopefully addressed most of these. Let me know if you see anything else that needs cleaning up. I tested all these changes with my experimental driver and the output was the correct/as expected.

Also, instead of trying to have a limited number of interrupt queue anchors and rounding submitted interrupt transfers to the closest existing 2^n interval interrupt anchor queue based upon the transfer's requested ms_interval (which is how XHCI recommends the host controller driver handle low/full speed interrupt transfers and how all high speed+ interrupt transfers are handled in general), I think it would be better to link them all into the schedule so that they are executed by the hardware controller every 1ms frame. That way we don't have two sources of jitter (one from interrupt transfer "misses" when the value is not in a queue that corresponds to it's actual requested interval time, and another from the polling thread introduced here which will inevitably go out of sync from the hardware), and another annoying bug I was dealing with is addressed.

The only downside to this is extra bandwidth usage, but that doesn't matter to us currently because 1.) we don't actually support any USB devices other than hubs so we aren't exactly hurting for bandwidth, and 2.) the UHCI schedule is executed breadth-first as we have it configured, so any extra bandwidth usage per frame is minimal, and finally 3.) if a device doesn't have any data to send, it's not an error or exceptional condition, it will just send a NAK. I have this new approach added here and tested it as well with my USB keyboard "driver" and it worked fine.

**Review by ADKaster — APPROVED** [review]

**Pull request merged**
